libraries/common/stream.effekt

```diff
@@ -0,0 +1,417 @@
+      } else {
+        do stop()
+      }
+    }
+  }
+}
+
+def feed(bytes: ByteArray) { reader: () => Unit / read[Byte] } = {
+  var i = 0
+  try {
+    reader()
+  } with read[Byte] { () =>
+    resume {
+      if (i < bytes.size) {
+        bytes.unsafeGet(i)
+      } else {
+        do stop()
+      }
+    }
+  }
+}
+
+def source[A, R] { stream: () => Unit / emit[A] } { reader: () => R / read[A] }: R = {
+  var next = box { () => None() }
+  next = box { () =>
+    try {
+      stream()
+      None()
+    } with emit[A] {
+      def emit(v) = {
+        next = box { () => resume(()) }
+        Some(v)
+      }
+    }
+  }
+  try {
+    reader()
+  } with read[A] { () =>
+    resume {
+      next().getOrElse { do stop() }
+    }
+  }
+}
+
+
+def writeFile[R](path: String) { stream: () => R / emit[Byte] }: R / Exception[IOError] = {
+
+  val file = open(path, WriteOnly());
+  with on[IOError].finalize { close(file) }
+
+  val chunkSize = 1048576 // 1MB
+  val buffer = bytearray::allocate(chunkSize)
+  var offset = 0
+
+  def push(i: Int, n: Int): Unit = {
+    val r = write(file, buffer, i, n, -1)
+    if (r < n) {
+      push(i + r, n - r)
+    }
+  }
+
+  try {
+    val r = stream()
+    push(0, offset)
+    return r
+  } with emit[Byte] {
+    def emit(byte) = {
+      if (offset >= buffer.size) { push(0, buffer.size) }
+      buffer.unsafeSet(offset, byte)
+      offset = offset + 1
+      resume(())
+    }
+  }
+}
+
+def readFile[R](path: String) { reader: () => R / read[Byte] }: R / Exception[IOError] = {
+
+  val file = open(path, ReadOnly());
+  with on[IOError].finalize { close(file) }
+
+  val chunkSize = 1048576 // 1MB
+  val buffer = bytearray::allocate(chunkSize)
+  var offset = 0
+  var length = 0
+
+  def pull(): Unit / stop = {
+    read(file, buffer, 0, chunkSize, -1) match {
+      case 0 =>
+        do stop()
+      case n =>
+        offset = 0
+        length = n
+    }
+  }
+
+  try {
+    reader()
+  } with read[Byte] { () =>
+    resume {
+      if (offset >= length) { pull() }
+      val byte = buffer.unsafeGet(offset)
+      offset = offset + 1
+      return byte
+    }
+  }
+}
+
+def decodeChar(): Char / {read[Byte], stop} = {
+  val b = do read().toInt
+  if (b < 128) {
+    b.toChar
+  } else if (b < 224) {
+    val part1 = bitwiseShl(bitwiseAnd(b, 31), 6)
+    val part2 = bitwiseAnd(do read().toInt, 63)
+    bitwiseOr(part1, part2).toChar
+  } else if (b < 240) {
+    val part1 = bitwiseShl(bitwiseAnd(b, 15), 12)
+    val part2 = bitwiseShl(bitwiseAnd(do read().toInt, 63), 6)
+    val part3 = bitwiseAnd(do read().toInt, 63)
+    bitwiseOr(bitwiseOr(part1, part2), part3).toChar
+  } else {
+    val part1 = bitwiseShl(bitwiseAnd(b, 7), 18)
+    val part2 = bitwiseShl(bitwiseAnd(do read().toInt, 63), 12)
+    val part3 = bitwiseShl(bitwiseAnd(do read().toInt, 63), 6)
+    val part4 = bitwiseAnd(do read().toInt, 63)
+    bitwiseOr(bitwiseOr(bitwiseOr(part1, part2), part3), part4).toChar
+  }
+}
```

Comment:
Why do you mix `do read().toInt` and the decoding? Shouldn't the decoding go into strings or char or whatever?

Reply:
We talked about this already... it can't be do to the character wise streaming `do read()`